contracts/keepers/FliRebalanceKeeper.sol

```diff
@@ -0,0 +1,79 @@
+/*
+    Copyright 2021 Index Cooperative.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity 0.6.10;
+pragma experimental ABIEncoderV2;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { KeeperCompatibleInterface } from "@chainlink/contracts/src/v0.6/KeeperCompatible.sol";
+
+/**
+ * @title RebalanceKeeper
+ * @author Index Cooperative
+ * 
+ * Chainlink Keeper which automatically rebalances FLI SetTokens.
+ */
+contract FliRebalanceKeeper is KeeperCompatibleInterface {
+    using Address for address;
+
+    /* ============ Modifiers ============ */
+    modifier onlyRegistry() {
+        require(msg.sender == registryAddress, "Only registry address can call this function");
+        _;
+    }
+
+    /* ============ State Variables ============ */
+
+    address public fliExtension;                          // Address of the fli extension contract
+    address public registryAddress;                       // Address of the chainlink keeper registry
+
+    /* ============ Constructor ============ */
+    constructor(address _fliExtension, address _registryAddress) public {
+        fliExtension = _fliExtension;
+        registryAddress = _registryAddress;
+    }    
+
+    /**
+     * As checkUpkeep is not a view function, calling this function will actually consume gas.
+     * As such if a keeper calls this function, it will always return true so that performUpkeep will be called.
+     */    
+    function checkUpkeep(bytes calldata /* checkData */) external override onlyRegistry returns (bool, bytes memory) {
+        return (true, new bytes(0));
+    }
+
+    /**
+     * performUpkeep checks that a rebalance is required. Otherwise the contract call will revert.
+     */
+    function performUpkeep(bytes calldata /* performData */) external override onlyRegistry {
+        Address.functionCall(fliExtension, getRebalanceCalldata());
+    }
+
+    function getRebalanceCalldata() private returns (bytes memory) {
+        bytes memory shouldRebalanceCalldata = abi.encodeWithSignature("shouldRebalance()");
+        bytes memory shouldRebalanceResponse = Address.functionCall(address(fliExtension), shouldRebalanceCalldata, "Failed to execute shouldRebalance()");
+        (string[] memory exchangeNames, uint256[] memory shouldRebalances) = abi.decode(shouldRebalanceResponse, (string[], uint256[]));
+
+        uint256 shouldRebalance = shouldRebalances[0];
```

Comment:
We might want to be able to select what exchange to use. Maybe we can store what index to use somewhere?

Reply:
Shouldn't the fliExtension contract determine which exchange to use? Rather than this contract? Or are you suggesting to have a whitelist of exchanges that this contract can call?

Reply:
At the moment we allow the caller of the FLI extension to determine what exchange to use (of a list of whitelisted exchanges). This is because determining the optimal exchange may be difficult or gas extensive to do on-chain. Lately we have just been hardcoding our bots to use the exchange we want, with a backup exchange already whitelisted in case we need to make changes quickly. For this contract since we have a bit less trust in the chainlink keepers, we probably want to hard code which exchange to use.

Reply:
Hmmm - then we keep a list of allowable addresses that we can contact? What's the exchangeName that comes back from the fliExtension call - I'm assuming it's an address?

Reply:
The exchange name is actually just a string which corresponds to the adapter that is used to do the trade part of the rebalance. I think its probably best to hardcode the index of the array that we want to use (and allow the owner of the contract to change it).